src/osa/scripts/gain_selection.py

```diff
@@ -263,64 +371,90 @@ def GainSel_finished(date: str) -> bool:
     return flagfile.exists()
 
 
-def check_failed_jobs(date: str, output_basedir: Path = None):
+def check_gainsel_jobs_runwise(date: str, run_id: int) -> bool:
     """Search for failed jobs in the log directory."""
-    failed_jobs = []
-    log_dir = output_basedir / "log" / date
-    filenames = glob.glob(f"{log_dir}/gain_selection*.log")
-    jobs = [re.search(r'(?<=_)(.[0-9.]+?)(?=.log)', i).group(0) for i in filenames]
+    base_dir = Path(cfg.get("LST1", "BASE"))
+    log_dir = base_dir / f"R0G/log/{date}"
+    history_files = log_dir.glob(f"gain_selection_{run_id:05d}.????.history")
+    failed_subruns = []
+    log.info(f"Checking all history files of run {run_id}")
+    
+    for file in list(history_files):
+        match = re.search(f"gain_selection_{run_id:05d}.(\d+).history", str(file))
+        subrun = match.group(1)
+        if file.read_text() != "":
+            gainsel_rc = file.read_text().splitlines()[-1][-1]
+
+            if gainsel_rc == "1":
+                log.warning(f"Gain selection failed for run {run_id}.{subrun}")
+                failed_subruns.append(file)
+
+            elif gainsel_rc == "0":
+                log.debug(f"Gain selection finished successfully for run {run_id}.{subrun}")
+        else:
+            log.info(f"Gain selection is still running for run {run_id}.{subrun}")
+ 
+    if failed_subruns:
```

Comment:
Maybe worth logging these failed_subruns as well?

Reply:
what do you mean? there are logging messages after this line

Reply:
yes, but you only mention some failed jobs for a given run. Isn't it useful to know which subruns as well?

Reply:
yes, but the subrun number of the failed job is also said above (in line 394)

Reply:
ok!

Reply:
One more comment: I was thinking about building a web page to watch the gain_selection operation, as we do for the sequencer. This failed_runs list could be used to compute the percentage of advance and build it.

Reply:
Good idea. Maybe the check option should export a table object that can be easily formatted and uploaded to the web with this information